README.md
# FanboyKit

Search iTunes with FanboyKit. This Swift package provides a client for [fanboy-http](https://github.com/michaelnisi/fanboy-http), a caching proxy for the [iTunes Search API](https://affiliate.itunes.apple.com/resources/documentation/itunes-store-web-service-search-api/). FanboyKit is used in the [Podest](https://github.com/michaelnisi/podest) podcast app.

## Example

Querying for suggestions matching the term `"crook"` limiting the result to 10.

```swift
import Foundation
import Patron
import FanboyKit

let url = URL(string: "https://your.endpoint")!
let s = URLSession(configuration: .default)
let p = Patron(URL: url, session: s)
let svc = Fanboy(client: p)

try! svc.suggestions(matching: "crook", limit: 10) { result, error in
  print(error ?? result)
}
```

Please refer to [fanboy-http](https://github.com/michaelnisi/fanboy-http) for details.

## Types

### FanboyError

```swift
enum FanboyError: Error {
  case unexpectedResult(result: AnyObject?)
  case cancelledByUser
  case invalidTerm
}
```

### FanboyService

```swift
protocol FanboyService {
  var client: JSONService { get }

  @discardableResult func version(
    completionHandler cb: @escaping (_ version: String?, Error?) -> Void
  ) -> URLSessionTask

  @discardableResult func search(
    term: String,
    completionHandler cb: @escaping (
      _ podcasts: [[String : AnyObject]]?, _ error: Error?) -> Void
  ) throws -> URLSessionTask

  @discardableResult func lookup(
    guids: [String],
    completionHandler cb: @escaping (
      _ podcasts: [[String : AnyObject]]?, _ error: Error?) -> Void
  ) -> URLSessionTask

  @discardableResult func suggestions(
    matching: String,
    limit: Int,
    completionHandler cb: @escaping (
      _ terms: [String]?, _ error: Error?) -> Void
  ) throws -> URLSessionTask
}
```

#### client

```swift
var client: JSONService { get }
```

The client property of the `FanboyService` object gives access to the underlying [Patron](https://github.com/michaelnisi/patron) client, providing hostname and status of the remote service.

## Test

With **fanboy-http** running, do:

```
$ swift test
```

## Install

📦 Add `https://github.com/michaelnisi/fanboy-kit`  to your package manifest.

## License

[MIT License](https://github.com/michaelnisi/fanboy-kit/blob/master/LICENSE)


CHANGELOG.md
## 9.0.0 (2020-11-22)

Convert to Swift Package.


Tests/LinuxMain.swift
import XCTest

import FanboyKitTests

var tests = [XCTestCaseEntry]()
tests += FanboyKitTests.allTests()
XCTMain(tests)


Tests/FanboyKitTests/FanboyKitTests.swift
//
//  FanboyKitTests.swift
//  FanboyKitTests
//
//  Created by Michael Nisi on 22/11/15.
//  Copyright © 2015 Michael Nisi. All rights reserved.
//

import XCTest
import Patron
@testable import FanboyKit

private func delay(_ ms: Int64 = Int64(arc4random_uniform(10)), cb: @escaping () -> Void) {
  let delta = ms * Int64(NSEC_PER_MSEC)
  let when = DispatchTime.now() + Double(delta) / Double(NSEC_PER_SEC)
  DispatchQueue.main.asyncAfter(deadline: when, execute: cb)
}

private func freshSession() -> URLSession {
  let conf = URLSessionConfiguration.default
  conf.httpShouldUsePipelining = false
  conf.requestCachePolicy = .reloadIgnoringLocalCacheData
  return URLSession(configuration: conf)
}

private func freshFanboy(_ url: NSURL) -> Fanboy {
  let session = freshSession()
  let client = Patron(URL: url as URL, session: session)
  return Fanboy(client: client)
}

final class InternalTests: XCTestCase {
  func testEncodeTerm() {
    let wanted = [
      "abc",
      "abc",
      "abc",
      "abc%20def",
      "abc%20%20def" // Remote service is taking care of inner whitespace.
    ]
    let found = [
      "abc",
      " abc",
      " abc ",
      " abc def ",
      " abc  def"
      ].map {
      try! encodeTerm($0)
    }
    for (i, wantedTerm) in wanted.enumerated() {
      let foundTerm = found[i]
      XCTAssertEqual(foundTerm, wantedTerm)
    }
  }
}

final class FanboyFailureTests: XCTestCase {

  var svc: FanboyService!

  override func setUp() {
    super.setUp()

    let url = URL(string: "http://localhost:8385")!
    svc = freshFanboy(url as NSURL)
  }

  override func tearDown() {
    super.tearDown()
  }

  func callbackWithExpression (_ exp: XCTestExpectation) -> (Any?, Error?) -> Void {
    func cb (result: Any?, error: Error?) -> Void {
      let er = error! as NSError
      XCTAssertEqual(er.code, -1004)
      XCTAssertNil(result)

      let (code, _) = svc.client.status!
      XCTAssertEqual(code, er.code)

      exp.fulfill()
    }
    return cb
  }

  func testSuggest() {
    let exp = self.expectation(description: "suggest")
    let cb = callbackWithExpression(exp)
    try! svc.suggestions(matching: "f", limit: 10, completionHandler: cb)
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  func testLookup() {
    let exp = self.expectation(description: "lookup")
    let cb = callbackWithExpression(exp)
    svc.lookup(guids: ["528458508", "974240842"], completionHandler: cb)
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  func testSearch() {
    let exp = self.expectation(description: "search")
    let cb = callbackWithExpression(exp)
    try! svc.search(term: "fireball", completionHandler: cb)
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  func testVersion() {
    let exp = self.expectation(description: "version")
    let cb = callbackWithExpression(exp)
    svc.version(completionHandler: cb)
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }
}

class FanboySuccessTests: XCTestCase {

  var svc: FanboyService!

  override func setUp() {
    super.setUp()

    let url = URL(string: "http://localhost:8383")!
    svc = freshFanboy(url as NSURL)
  }

  override func tearDown() {
    svc = nil
    super.tearDown()
  }

  func testHost() {
    XCTAssertEqual(svc.client.host, "localhost")
  }

  // MARK: Searching

  func testSearch() {
    let names = self.names
    let exp = self.expectation(description: "search")
    let term = "fireball"
    try! svc.search(term: term) { feeds, error in
      XCTAssertNil(error)
      feeds!.forEach() { feed in
        names.forEach() { name in
          XCTAssertNotNil(feed[name])
        }
      }
      exp.fulfill()
    }
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  func testSearchtWithInvalidQuery() {
    let queries = ["", " "]
    var count = 0
    queries.forEach() { query in
      do {
        try svc.search(term: query) { _, _ in }
      } catch FanboyError.invalidTerm {
        count += 1
      } catch {
        XCTFail("should not throw unexpected error")
      }
    }
    XCTAssertEqual(count, queries.count)
  }

  func testSearchCancel () {
    let svc = self.svc!
    let exp = self.expectation(description: "search")
    let op = try! svc.search(term: "fireball") { feeds, error in
      defer {
        exp.fulfill()
      }
      guard feeds == nil else {
        return
      }
      do {
        throw error!
      } catch FanboyError.cancelledByUser {
      } catch {
        XCTFail("should be expected error")
      }
    }
    delay() {
      op.cancel()
    }
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  // MARK: Suggesting

  func testSuggest() {
    let exp = self.expectation(description: "suggest")
    func next() {
      try! svc.suggestions(matching: "f", limit: 10) { terms, error in
        XCTAssertNil(error)
        XCTAssert(terms!.contains("fireball"))
        exp.fulfill()
      }
    }
    try! svc.search(term: "fireball") { feeds, error in
      XCTAssertNil(error)
      next()
    }
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  func testSuggestWithInvalidQuery() {
    let queries = ["", " "]
    var count = 0
    queries.forEach() { q in
      do {
        try svc.suggestions(matching: q, limit: 10) { _, _ in }
      } catch FanboyError.invalidTerm {
        count += 1
      } catch {
        XCTFail("should not throw unexpected error")
      }
    }
    XCTAssertEqual(count, queries.count)
  }

  func testSuggestCancel () {
    let svc = self.svc!
    let exp = self.expectation(description: "suggest")
    let term = "f"
    try! svc.suggestions(matching: term, limit: 10) { terms, error in
      do {
        throw error!
      } catch FanboyError.cancelledByUser {
      } catch {
        XCTFail("should be expected error")
      }
      XCTAssertNil(terms)
      exp.fulfill()
      }.cancel()
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  // MARK: Uplooking

  let names = ["author", "title", "img100", "guid", "img30", "img60", "img600", "updated"]

  func testLookup() {
    let exp = self.expectation(description: "lookup")
    let names = self.names
    let guids = ["528458508", "974240842"]

    // These get cached rather agressively.

    svc.lookup(guids: guids) { feeds, error in
      XCTAssertNil(error)
      XCTAssertEqual(feeds!.count, 2)
      feeds!.forEach() { feed in
        names.forEach() { name in
          XCTAssertNotNil(feed[name])
        }
      }
      exp.fulfill()
    }
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  func testLookupCancel() {
    let svc = self.svc!
    let exp = self.expectation(description: "lookup")
    let guids = ["528458508", "974240842"]
    let op = svc.lookup(guids: guids) { feeds, error in
      defer {
        exp.fulfill()
      }
      guard feeds == nil else {
        return
      }
      do {
        throw error!
      } catch FanboyError.cancelledByUser {
      } catch {
        XCTFail("should be expected error")
      }
    }
    delay() {
      op.cancel()
    }
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  // MARK: Version

  func testVersion() {
    let exp = self.expectation(description: "version")
    svc.version { version, error in
      XCTAssertNil(error)
      XCTAssertEqual(version, "5.0.0")
      exp.fulfill()
    }
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  func testVersionCancel() {
    let svc = self.svc!
    let exp = self.expectation(description: "version")
    let op = svc.version { version, error in
      do {
        throw error!
      } catch FanboyError.cancelledByUser {
      } catch {
        XCTFail("should not be unexpected error")
      }
      XCTAssertNil(version)
      exp.fulfill()
    }
    op.cancel()
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }
}


Tests/FanboyKitTests/XCTestManifests.swift
import XCTest

#if !canImport(ObjectiveC)
public func allTests() -> [XCTestCaseEntry] {
    return [
        testCase(FanboyKitTests.allTests),
    ]
}
#endif


FanboyKit/index.swift
//
//  index.swift
//  FanboyKit
//
//  Created by Michael Nisi on 22/11/15.
//  Copyright © 2015 Michael Nisi. All rights reserved.
//

import Foundation
import Patron

// MARK: API

public enum FanboyError: Error {
  case unexpectedResult(result: AnyObject?)
  case cancelledByUser
  case invalidTerm
}

/// A client for the fanboy-http, a caching proxy of the iTunes Search API.
public protocol FanboyService {
  var client: JSONService { get }

  @discardableResult func version(
    completionHandler cb: @escaping (_ version: String?, Error?) -> Void
  ) -> URLSessionTask

  @discardableResult func search(
    term: String,
    completionHandler cb: @escaping (_ podcasts: [[String : AnyObject]]?, _ error: Error?) -> Void
  ) throws -> URLSessionTask

  @discardableResult func lookup(
    guids: [String],
    completionHandler cb: @escaping (_ podcasts: [[String : AnyObject]]?, _ error: Error?) -> Void
  ) -> URLSessionTask

  @discardableResult func suggestions(
    matching: String,
    limit: Int,
    completionHandler cb: @escaping (_ terms: [String]?, _ error: Error?) -> Void
  ) throws -> URLSessionTask
}

// MARK: -

/// Transform errors.
private func retypeError(_ error: Error?) -> Error? {
  guard let er = error as NSError? else {
    return error
  }
  switch er.code {
  case -999: return FanboyError.cancelledByUser
  default: return er
  }
}

/// Removes whitespace at the beginning and end of the specified term, and
/// returns URL encoded term. Note that inner whitespace is left unmodified.
///
/// - parameter term: The term to encode.
///
/// - returns: URL encoded search term.
///
/// - throws: `FanboyError.InvalidTerm`
func encodeTerm(_ term: String) throws -> String {
  let ws = CharacterSet.whitespaces
  let trimmed = term.trimmingCharacters(in: ws)
  guard !trimmed.isEmpty else {
    throw FanboyError.invalidTerm
  }
  let url = CharacterSet.urlHostAllowed
  return trimmed.addingPercentEncoding(withAllowedCharacters: url)!
}

public final class Fanboy: FanboyService {

  /// The underlying JSON service client.
  public let client: JSONService

  /// Creates a `Fanboy` object with the specified client.
  ///
  /// - parameter client: The remote service to use.
  public init(client: JSONService) {
    self.client = client
  }

  private func request(
    _ path: String,
    cb: @escaping ([[String : AnyObject]]?, Error?) -> Void
  ) -> URLSessionTask {
    return client.get(path: path) { json, response, error in
      if let er = retypeError(error) {
        cb(nil, er)
      } else if let result = json as? [[String : AnyObject]] {
        cb(result, nil)
      } else {
        let er = FanboyError.unexpectedResult(result: json)
        cb(nil, er)
      }
    }
  }

  /// Lookup specific podcast feeds by their iTunes GUIDs.
  ///
  /// - parameter guids: The GUIDs of the podcasts to lookup in iTunes.
  /// - parameter completionHandler: A block with following parameters:
  /// - parameter error: An error object, or nil.
  /// - parameter podcasts: The podcasts with the requested `guids`.
  ///
  /// - returns: Returns the according URL session task.
  public func lookup(
    guids: [String],
    completionHandler cb: @escaping (_ podcasts: [[String : AnyObject]]?, _ error: Error?) -> Void
  ) -> URLSessionTask {
    let query = guids.joined(separator: ",")
    let path = "/lookup/\(query)"
    return request(path, cb: cb)
  }

  /// Search feeds matching the specified `term`.
  ///
  /// - parameter term: The search term, a space separated list of words, to
  /// search for.
  /// - parameter completionHandler: A block with following parameters:
  /// - parameter error: An error object, or nil.
  /// - parameter podcasts: The podcasts matching `term` in iTunes.
  ///
  /// - returns: Returns the according URL session task.
  public func search(
    term: String,
    completionHandler cb: @escaping (_ podcasts: [[String : AnyObject]]?, _ error: Error?) -> Void
  ) throws -> URLSessionTask {
    let t = try encodeTerm(term)
    let path = "/search?q=\(t)"
    return request(path, cb: cb)
  }

  /// Request suggestions for a given search term or fragment thereof.
  ///
  /// - parameter term: The term to find suggestions for, it has to be a single
  /// space separated list of lowercase words. But usually you'd pass fragments:
  /// leading characters of eventual search terms.
  /// - parameter limit: The maximum number of suggestions to receive.
  /// - parameter completionHandler: A block with following parameters:
  /// - parameter error: An error object, or nil.
  /// - parameter terms: The suggestions.
  ///
  /// - returns: Returns the according URL session task.
  ///
  /// - throws: Throws if `term ` could not be encoded to a valid search term.
  public func suggestions(
    matching term: String,
    limit: Int,
    completionHandler cb: @escaping (_ terms: [String]?, _ error: Error?) -> Void
  ) throws -> URLSessionTask {
    let t = try encodeTerm(term)
    let path = "/suggest?q=\(t)&max=\(limit)"
    return client.get(path: path) { json, response, error in
      if let er = retypeError(error) {
        cb(nil, er)
      } else if let result = json as? [String] {
        cb(result, nil)
      } else {
        let er = FanboyError.unexpectedResult(result: json)
        cb(nil, er)
      }
    }
  }

  /// Request the version of the remote service.
  ///
  /// - parameter completionHandler: A block with following parameters:
  /// - parameter error: An error object, or nil.
  /// - parameter version: The version of the remote service.
  ///
  /// - returns: Returns the according URL session task.
  public func version(
    completionHandler cb: @escaping (_ version: String?, _ error: Error?) -> Void
  ) -> URLSessionTask {
    return client.get(path: "/") { json, response, error in
      if let er = retypeError(error) {
        cb(nil, er)
      } else if let version = json?["version"] as? String {
        cb(version, nil)
      } else {
        let er = FanboyError.unexpectedResult(result: json)
        cb(nil, er)
      }
    }
  }
}


Sources/FanboyKit/FanboyKit.swift
//
//  index.swift
//  FanboyKit
//
//  Created by Michael Nisi on 22/11/15.
//  Copyright © 2015 Michael Nisi. All rights reserved.
//

import Foundation
import Patron

// MARK: API

public enum FanboyError: Error {
  case unexpectedResult(result: AnyObject?)
  case cancelledByUser
  case invalidTerm
}

/// A client for the fanboy-http, a caching proxy of the iTunes Search API.
public protocol FanboyService {
  var client: JSONService { get }

  @discardableResult func version(
    completionHandler cb: @escaping (_ version: String?, Error?) -> Void
  ) -> URLSessionTask

  @discardableResult func search(
    term: String,
    completionHandler cb: @escaping (
      _ podcasts: [[String : AnyObject]]?, _ error: Error?) -> Void
  ) throws -> URLSessionTask

  @discardableResult func lookup(
    guids: [String],
    completionHandler cb: @escaping (
      _ podcasts: [[String : AnyObject]]?, _ error: Error?) -> Void
  ) -> URLSessionTask

  @discardableResult func suggestions(
    matching: String,
    limit: Int,
    completionHandler cb: @escaping (
      _ terms: [String]?, _ error: Error?) -> Void
  ) throws -> URLSessionTask
}

/// Transform errors.
fileprivate func retypeError(_ error: Error?) -> Error? {
  guard let er = error as NSError? else {
    return error
  }
  switch er.code {
  case -999: return FanboyError.cancelledByUser
  default: return er
  }
}

fileprivate func escapeDoubleQuotes(in string: String) -> String {
  return string.replacingOccurrences(
    of: "\"",
    with: "\\\"",
    options: String.CompareOptions.literal,
    range: nil
  )
}

/// Removes whitespace at the beginning and end of the specified term, and
/// returns URL encoded term. Note that inner whitespace is left unmodified.
///
/// - Parameter term: The term to encode.
///
/// - Returns: URL encoded search term.
///
/// - Throws: `FanboyError.InvalidTerm`
func encodeTerm(_ term: String) throws -> String {
  let ws = CharacterSet.whitespaces
  let a = term.trimmingCharacters(in: ws)
  guard !a.isEmpty else {
    throw FanboyError.invalidTerm
  }
  
  let b = escapeDoubleQuotes(in: a)
  
  let urlHostAllowed = CharacterSet.urlHostAllowed
  assert(!urlHostAllowed.contains("\""))
  
  return b.addingPercentEncoding(withAllowedCharacters: urlHostAllowed)!
}


public final class Fanboy: FanboyService {

  /// The underlying JSON service client.
  public let client: JSONService

  /// Creates a `Fanboy` object with the specified client.
  ///
  /// - Parameter client: The remote service to use.
  public init(client: JSONService) {
    self.client = client
  }

  private func request(
    _ path: String,
    cb: @escaping ([[String : AnyObject]]?, Error?) -> Void
  ) -> URLSessionTask {
    return client.get(path: path) { json, response, error in
      if let er = retypeError(error) {
        cb(nil, er)
      } else if let result = json as? [[String : AnyObject]] {
        cb(result, nil)
      } else {
        let er = FanboyError.unexpectedResult(result: json)
        cb(nil, er)
      }
    }
  }

  /// Lookup specific podcast feeds by their iTunes GUIDs.
  ///
  /// - Parameters:
  ///   - guids: The GUIDs of the podcasts to lookup in iTunes.
  ///   - completionHandler: A block with following parameters:
  ///   - error: An error object, or nil.
  ///   - podcasts: The podcasts with the requested `guids`.
  ///
  /// - Returns: Returns the according URL session task.
  public func lookup(
    guids: [String],
    completionHandler cb: @escaping (
    _ podcasts: [[String : AnyObject]]?, _ error: Error?) -> Void
  ) -> URLSessionTask {
    let query = guids.joined(separator: ",")
    let path = "/lookup/\(query)"
    return request(path, cb: cb)
  }

  /// Search feeds matching the specified `term`.
  ///
  /// - Parameters:
  ///   - term: The search term, a space separated list of words, to
  /// search for.
  ///   - completionHandler: A block with following parameters:
  ///   - error: An error object, or nil.
  ///   - podcasts: The podcasts matching `term` in iTunes.
  ///
  /// - Returns: Returns the according URL session task.
  public func search(
    term: String,
    completionHandler cb: @escaping (
    _ podcasts: [[String : AnyObject]]?, _ error: Error?) -> Void
  ) throws -> URLSessionTask {
    let t = try encodeTerm(term)
    let path = "/search?q=\(t)&max=50"
    return request(path, cb: cb)
  }

  /// Request suggestions for a given search term or fragment thereof.
  ///
  /// - Parameters:
  ///   - term: The term to find suggestions for, it has to be a single
  /// space separated list of lowercase words. But usually you'd pass fragments:
  /// leading characters of eventual search terms.
  ///   - limit: The maximum number of suggestions to receive.
  ///   - completionHandler: A block with following parameters:
  ///   - error: An error object, or nil.
  ///   - terms: The suggestions.
  ///
  /// - Returns: Returns the according URL session task.
  ///
  /// - Throws: Throws if `term ` could not be encoded to a valid search term.
  public func suggestions(
    matching term: String,
    limit: Int,
    completionHandler cb: @escaping (
    _ terms: [String]?, _ error: Error?) -> Void
  ) throws -> URLSessionTask {
    let t = try encodeTerm(term)
    let path = "/suggest?q=\(t)&max=10"
    return client.get(path: path) { json, response, error in
      if let er = retypeError(error) {
        cb(nil, er)
      } else if let result = json as? [String] {
        cb(result, nil)
      } else {
        let er = FanboyError.unexpectedResult(result: json)
        cb(nil, er)
      }
    }
  }

  /// Request the version of the remote service.
  ///
  /// - Parameters:
  ///   - completionHandler: A block with following parameters:
  ///   - error: An error object, or nil.
  ///   - version: The version of the remote service.
  ///
  /// - Returns: Returns the according URL session task.
  public func version(
    completionHandler cb: @escaping (
    _ version: String?, _ error: Error?) -> Void
  ) -> URLSessionTask {
    return client.get(path: "/") { json, response, error in
      if let er = retypeError(error) {
        cb(nil, er)
      } else if let version = json?["version"] as? String {
        cb(version, nil)
      } else {
        let er = FanboyError.unexpectedResult(result: json)
        cb(nil, er)
      }
    }
  }
}


FanboyKitTests/FanboyKitTests.swift
//
//  FanboyKitTests.swift
//  FanboyKitTests
//
//  Created by Michael Nisi on 22/11/15.
//  Copyright © 2015 Michael Nisi. All rights reserved.
//

import XCTest
import Patron
@testable import FanboyKit

private func delay(_ ms: Int64 = Int64(arc4random_uniform(10)), cb: @escaping () -> Void) {
  let delta = ms * Int64(NSEC_PER_MSEC)
  let when = DispatchTime.now() + Double(delta) / Double(NSEC_PER_SEC)
  DispatchQueue.main.asyncAfter(deadline: when, execute: cb)
}

private func freshSession() -> URLSession {
  let conf = URLSessionConfiguration.default
  conf.httpShouldUsePipelining = false
  conf.requestCachePolicy = .reloadIgnoringLocalCacheData
  return URLSession(configuration: conf)
}

private func freshFanboy(_ url: NSURL) -> Fanboy {
  let session = freshSession()
  let client = Patron(URL: url as URL, session: session)
  return Fanboy(client: client)
}

final class InternalTests: XCTestCase {
  func testEncodeTerm() {
    let wanted = [
      "abc",
      "abc",
      "abc",
      "abc%20def",
      "abc%20%20def" // Remote service is taking care of inner whitespace.
    ]
    let found = [
      "abc",
      " abc",
      " abc ",
      " abc def ",
      " abc  def"
      ].map {
      try! encodeTerm($0)
    }
    for (i, wantedTerm) in wanted.enumerated() {
      let foundTerm = found[i]
      XCTAssertEqual(foundTerm, wantedTerm)
    }
  }
}

final class FanboyFailureTests: XCTestCase {

  var svc: FanboyService!

  override func setUp() {
    super.setUp()

    let url = URL(string: "http://localhost:8385")!
    svc = freshFanboy(url as NSURL)
  }

  override func tearDown() {
    super.tearDown()
  }

  func callbackWithExpression (_ exp: XCTestExpectation) -> (Any?, Error?) -> Void {
    func cb (result: Any?, error: Error?) -> Void {
      let er = error! as NSError
      XCTAssertEqual(er.code, -1004)
      XCTAssertNil(result)

      let (code, _) = svc.client.status!
      XCTAssertEqual(code, er.code)

      exp.fulfill()
    }
    return cb
  }

  func testSuggest() {
    let exp = self.expectation(description: "suggest")
    let cb = callbackWithExpression(exp)
    try! svc.suggestions(matching: "f", limit: 10, completionHandler: cb)
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  func testLookup() {
    let exp = self.expectation(description: "lookup")
    let cb = callbackWithExpression(exp)
    svc.lookup(guids: ["528458508", "974240842"], completionHandler: cb)
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  func testSearch() {
    let exp = self.expectation(description: "search")
    let cb = callbackWithExpression(exp)
    try! svc.search(term: "fireball", completionHandler: cb)
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  func testVersion() {
    let exp = self.expectation(description: "version")
    let cb = callbackWithExpression(exp)
    svc.version(completionHandler: cb)
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }
}

class FanboySuccessTests: XCTestCase {

  var svc: FanboyService!

  override func setUp() {
    super.setUp()

    let url = URL(string: "http://localhost:8383")!
    svc = freshFanboy(url as NSURL)
  }

  override func tearDown() {
    svc = nil
    super.tearDown()
  }

  func testHost() {
    XCTAssertEqual(svc.client.host, "localhost")
  }

  // MARK: Searching

  func testSearch() {
    let names = self.names
    let exp = self.expectation(description: "search")
    let term = "fireball"
    try! svc.search(term: term) { feeds, error in
      XCTAssertNil(error)
      feeds!.forEach() { feed in
        names.forEach() { name in
          XCTAssertNotNil(feed[name])
        }
      }
      exp.fulfill()
    }
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  func testSearchtWithInvalidQuery() {
    let queries = ["", " "]
    var count = 0
    queries.forEach() { query in
      do {
        try svc.search(term: query) { _, _ in }
      } catch FanboyError.invalidTerm {
        count += 1
      } catch {
        XCTFail("should not throw unexpected error")
      }
    }
    XCTAssertEqual(count, queries.count)
  }

  func testSearchCancel () {
    let svc = self.svc!
    let exp = self.expectation(description: "search")
    let op = try! svc.search(term: "fireball") { feeds, error in
      defer {
        exp.fulfill()
      }
      guard feeds == nil else {
        return
      }
      do {
        throw error!
      } catch FanboyError.cancelledByUser {
      } catch {
        XCTFail("should be expected error")
      }
    }
    delay() {
      op.cancel()
    }
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  // MARK: Suggesting

  func testSuggest() {
    let exp = self.expectation(description: "suggest")
    func next() {
      try! svc.suggestions(matching: "f", limit: 10) { terms, error in
        XCTAssertNil(error)
        XCTAssert(terms!.contains("fireball"))
        exp.fulfill()
      }
    }
    try! svc.search(term: "fireball") { feeds, error in
      XCTAssertNil(error)
      next()
    }
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  func testSuggestWithInvalidQuery() {
    let queries = ["", " "]
    var count = 0
    queries.forEach() { q in
      do {
        try svc.suggestions(matching: q, limit: 10) { _, _ in }
      } catch FanboyError.invalidTerm {
        count += 1
      } catch {
        XCTFail("should not throw unexpected error")
      }
    }
    XCTAssertEqual(count, queries.count)
  }

  func testSuggestCancel () {
    let svc = self.svc!
    let exp = self.expectation(description: "suggest")
    let term = "f"
    try! svc.suggestions(matching: term, limit: 10) { terms, error in
      do {
        throw error!
      } catch FanboyError.cancelledByUser {
      } catch {
        XCTFail("should be expected error")
      }
      XCTAssertNil(terms)
      exp.fulfill()
      }.cancel()
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  // MARK: Uplooking

  let names = ["author", "title", "img100", "guid", "img30", "img60", "img600", "updated"]

  func testLookup() {
    let exp = self.expectation(description: "lookup")
    let names = self.names
    let guids = ["528458508", "974240842"]

    // These get cached rather agressively.

    svc.lookup(guids: guids) { feeds, error in
      XCTAssertNil(error)
      XCTAssertEqual(feeds!.count, 2)
      feeds!.forEach() { feed in
        names.forEach() { name in
          XCTAssertNotNil(feed[name])
        }
      }
      exp.fulfill()
    }
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  func testLookupCancel() {
    let svc = self.svc!
    let exp = self.expectation(description: "lookup")
    let guids = ["528458508", "974240842"]
    let op = svc.lookup(guids: guids) { feeds, error in
      defer {
        exp.fulfill()
      }
      guard feeds == nil else {
        return
      }
      do {
        throw error!
      } catch FanboyError.cancelledByUser {
      } catch {
        XCTFail("should be expected error")
      }
    }
    delay() {
      op.cancel()
    }
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  // MARK: Version

  func testVersion() {
    let exp = self.expectation(description: "version")
    svc.version { version, error in
      XCTAssertNil(error)
      XCTAssertEqual(version, "4.0.1")
      exp.fulfill()
    }
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }

  func testVersionCancel() {
    let svc = self.svc!
    let exp = self.expectation(description: "version")
    let op = svc.version { version, error in
      do {
        throw error!
      } catch FanboyError.cancelledByUser {
      } catch {
        XCTFail("should not be unexpected error")
      }
      XCTAssertNil(version)
      exp.fulfill()
    }
    op.cancel()
    self.waitForExpectations(timeout: 10) { er in
      XCTAssertNil(er)
    }
  }
}
